### ISSUE TITLE
Old map doesn't seem to be working with vector rendering until you republish them

### ISSUE THREAD

**Issue by saleiva** [CartoDB/cartodb#12333]
### Context
In some cases (sorry but such as this one https://team.carto.com/u/saleiva/builder/9aa6a817-ec2d-458a-b5c2-e1b4eae29db1 - it works now) maps render empty using the vector engine until you republish them.

It looks to be related with only quite old maps.

cc/ @juanignaciosl @xavijam

**Comment by xavijam** [comment]
@juanignaciosl will take a look, although our guessing is due to the cache.

**Comment by juanignaciosl** [comment]
Ok, after some testing, I might get it.

I reproduced it with [this map](https://juanignaciosl.carto.com/builder/e1f52b2a-8bc3-11e6-9814-0ef7f98ade21/embed). There's a difference in the vizjson between both (other than `updated_at`, of course): 

![captura de pantalla 2017-06-16 a las 8 53 55](https://user-images.githubusercontent.com/932968/27215416-5fb3962c-5271-11e7-806b-f798d32f2dc1.png)

As you see, the leftmost one (the one that doesn't work) lacks `simple_geom: point` option at the source analysis. The analysis has it at the database, so it seems that it lacked that attribute when it was published.

Since correlation doesn't imply causation, I'd like to get more information about this if possible before ruining other maps that reproduce the issue (I don't know anything about `simple_geom` but the obvious). Does this ring any bell to you? cc @alonsogarciapablo @donflopez 


.... (Ignore the following, they're notes for me)

Things that I've tested:
- Opening old Editor maps: work (expected, probably, as moving it to Builder forces the publishing).
- I've tested many of my maps.
  - [This one renders correctly, but there are 404s for many tiles](https://team.carto.com/u/juanignaciosl/builder/d298d60e-9eee-11e4-bc2e-0e0c41326911/embed). Asked Flopez.
  - [This one doesn't display data](https://juanignaciosl.carto.com/builder/4ceb00f0-200b-11e6-9468-0ea31932ec1d/embed), but it might be caused by the image markers. Asked Flopez.
  - Maps opened: 17.
  - Rendered with vector: 6.
  - Errors:
    - 3? (image markers): 4ceb00f0-200b-11e6-9468-0ea31932ec1d, e541adb0-204a-11e6-99ce-0e31c9be1b51, 5b8843c4-20c5-11e6-99ce-0e31c9be1b51.
    - 1? e1f52b2a-8bc3-11e6-9814-0ef7f98ade21

**Comment by juanignaciosl** [comment]
I've manually added `simple_geom: point` to the mapcap of the failing map and it works. So, the problem is the lack of that attribute, which is set by JS at frontend. Every visualization edited after Sept 12, 2016 should work, and the workaround is easy and safe. I think we're good. What do you think?

**Comment by donflopez** [comment]
@juanignaciosl I think that the problem is the layer name, old maps call them like layer0, layer1, layer2... but we look for the layer ID.

The 404 is the default response when a tile has no data, @dgaubert can tell us more.

**Comment by juanignaciosl** [comment]
I'll take a look at that. I just checked another working map and `simple_geom` doesn't seem to be the cause.

**Comment by juanignaciosl** [comment]
Ok, I reproduced it again with this map by @saleiva: https://team.carto.com/u/saleiva/builder/48449bca-5d54-11e6-9c31-0ee66e2c9693/embed

This time I began doing manually what backend does. It begins working when named map is republished.

# Differences

## Layers

Broken one layers:

```ruby
       {"type"=>"cartodb",
        "options"=>
         {"id"=>"edb880fd-c3cd-4227-9bb4-ab2d797f1ce6",
          "cartocss"=>
           "#layer {\n  polygon-fill: ramp([income_per_capita], (#d3f2a3, #97e196, #6cc08b, #4c9b82, #217a79, #105965, #074050), quantiles);\n  line-width: 0.6;\n  line-color: #122019;\n  line-opacity: 0.25;\n}",
          "cartocss_version"=>"2.1.1",
          "source"=>{"id"=>"e0"},
          "sql_wrap"=>nil}},
       {"type"=>"cartodb",
        "options"=>
         {"id"=>"4c8b2b2d-33e2-4d58-ada9-7810f19a4364",
          "cartocss"=>
           "#layer {\n  polygon-fill: #374c70;\n  polygon-opacity: 0;\n  line-width: 1.5;\n  line-color: #ffffff;\n  line-opacity: 1;\n}",
          "cartocss_version"=>"2.1.1",
          "source"=>{"id"=>"b1"},
          "sql_wrap"=>nil}},
       {"type"=>"cartodb",
        "options"=>
         {"id"=>"e7c98240-1249-4279-92e7-b5083418103f",
          "cartocss"=>
           "#layer {\n  marker-width: 6;\n  marker-fill: #ffffff;\n  marker-fill-opacity: 0.9;\n  marker-allow-overlap: true;\n  marker-line-width: 0.75;\n  marker-line-color: #000000;\n  marker-line-opacity: 1;\n}",
          "cartocss_version"=>"2.1.1",
          "source"=>{"id"=>"a0"},
          "sql_wrap"=>nil}}],
```

Fixed one layers:

```ruby
      {"id"=>"edb880fd-c3cd-4227-9bb4-ab2d797f1ce6",
        "type"=>"cartodb",
        "options"=>
         {"cartocss"=>
           "#layer {\n  polygon-fill: ramp([income_per_capita], (#d3f2a3, #97e196, #6cc08b, #4c9b82, #217a79, #105965, #074050), quantiles);\n  line-width: 0.6;\n  line-color: #122019;\n  line-opacity: 0.25;\n}",
          "cartocss_version"=>"2.1.1",
          "sql"=>
           "SELECT * FROM (SELECT * FROM sf_census_tracts_augmented) AS wrapped_query WHERE <%= layer0 %>=1",
          "sql_wrap"=>nil}},
       {"id"=>"4c8b2b2d-33e2-4d58-ada9-7810f19a4364",
        "type"=>"cartodb",
        "options"=>
         {"cartocss"=>
           "#layer {\n  polygon-fill: #374c70;\n  polygon-opacity: 0;\n  line-width: 1.5;\n  line-color: #ffffff;\n  line-opacity: 1;\n}",
          "cartocss_version"=>"2.1.1",
          "sql"=>
           "SELECT * FROM (SELECT * FROM sf_census_tracts_augmented) AS wrapped_query WHERE <%= layer1 %>=1",
          "sql_wrap"=>nil}},
       {"id"=>"e7c98240-1249-4279-92e7-b5083418103f",
        "type"=>"cartodb",
        "options"=>
         {"cartocss"=>
           "#layer {\n  marker-width: 6;\n  marker-fill: #ffffff;\n  marker-fill-opacity: 0.9;\n  marker-allow-overlap: true;\n  marker-line-width: 0.75;\n  marker-line-color: #000000;\n  marker-line-opacity: 1;\n}",
          "cartocss_version"=>"2.1.1",
          "sql"=>
           "SELECT * FROM (SELECT * FROM gyms_and_pokestops) AS wrapped_query WHERE <%= layer2 %>=1",
          "sql_wrap"=>nil}}],
```

- Broken one has the layer id at `options.id`, the new one is just `id`. That was changed [here](https://github.com/CartoDB/cartodb/commit/41e5aa403b763565ee65c46a37643b52b6905c90#diff-37de14bd6f97d298cf49816cf53dbf67). This is likely to be the issue.
- Broken one has `options.source`, new one has `options.sql`. That one puzzles me, because, looking at the code, I think that it should have `source`, not sql. Probably unrelated, though.

## Analyses

Broken one:

```ruby
      [{"id"=>"e0",
        "type"=>"source",
        "params"=>{"query"=>"SELECT * FROM sf_census_tracts_augmented"},
        "options"=>{"table_name"=>"sf_census_tracts_augmented"}},
       {"id"=>"b1",
        "type"=>"aggregate-intersection",
        "params"=>
         {"source"=>
           {"id"=>"e0",
            "type"=>"source",
            "params"=>{"query"=>"SELECT * FROM sf_census_tracts_augmented"},
            "options"=>{"table_name"=>"sf_census_tracts_augmented"}},
          "target"=>
           {"id"=>"a0",
            "type"=>"source",
            "params"=>{"query"=>"SELECT * FROM gyms_and_pokestops"},
            "options"=>{"table_name"=>"gyms_and_pokestops"}},
          "aggregate_function"=>"count",
          "aggregate_column"=>"cartodb_id"}},
       {"id"=>"a0",
        "type"=>"source",
        "params"=>{"query"=>"SELECT * FROM gyms_and_pokestops"},
        "options"=>{"table_name"=>"gyms_and_pokestops"}}]},
```

Fixed one:

```ruby
      [{"id"=>"e0",
        "type"=>"source",
        "params"=>{"query"=>"SELECT * FROM sf_census_tracts_augmented"}},
       {"id"=>"b1",
        "type"=>"aggregate-intersection",
        "params"=>
         {"source"=>
           {"id"=>"e0",
            "type"=>"source",
            "params"=>{"query"=>"SELECT * FROM sf_census_tracts_augmented"}},
          "target"=>
           {"id"=>"a0",
            "type"=>"source",
            "params"=>{"query"=>"SELECT * FROM gyms_and_pokestops"}},
          "aggregate_function"=>"count",
          "aggregate_column"=>"cartodb_id"}},
       {"id"=>"a0",
        "type"=>"source",
        "params"=>{"query"=>"SELECT * FROM gyms_and_pokestops"}}]},
```

- Fixed one doesn't have `options`.

### Preview layers

Broken one lacks `preview_layers`:

```ruby
     :preview_layers=>
      {:"edb880fd-c3cd-4227-9bb4-ab2d797f1ce6"=>true,
       :"4c8b2b2d-33e2-4d58-ada9-7810f19a4364"=>true,
       :"e7c98240-1249-4279-92e7-b5083418103f"=>true}}}}
```


# Specific workaround

_You can also just publish the map instead of doing this_.

To check if a visualization is affected, you can check this:

```ruby
Carto::NamedMaps::Api.new(Carto::Visualization.find('05e51f96-519d-11e7-a51d-784f436d1355')).show[:template][:layergroup][:layers][1]['id']
Carto::NamedMaps::Api.new(Carto::Visualization.find('05e51f96-519d-11e7-a51d-784f436d1355')).show[:template][:layergroup][:layers][1]['options']['id']
```

If first one doesn't display an id but the second does, it's broken because of the named map template.

How to regenerate a named map:

```ruby
Carto::Visualization.find('<uuid>').send(:invalidation_service).send(:update_or_destroy_named_map)
```

@saleiva This only happens because of a bug fixed in November (#10775, [comment](https://github.com/CartoDB/cartodb/pull/10775#issuecomment-262288142)). As it's a quite old bug I think that we can just use the workaround for broken visualizations. The alternative is find broken visualizations and regenerate the named map.

**Comment by juanignaciosl** [comment]
I think that we can skip reviewing old visualizations, as this bug was fixed in november and there's a quick workaround. Please reopen if you don't agree.
